### ISSUE TITLE
Tsgo considers wrong API usage as correct - difference in typechecking

### ISSUE THREAD

**Issue by jansedlon** [microsoft/typescript-go#2889]
## Steps to reproduce
1. Clone repo https://github.com/jansedlon/tsgo-repro
2. Install dependencies
3. Open index.ts

## Behavior with `typescript@5.9`
The constant `posts` has a type of 
```ts
const posts: {
    id: string;
    createdAt: Date;
    updatedAt: Date;
}[]
```

And typescript reports error in `attachments` object because selecting database fields needs to be inside of `columns` object and `orderBy` in `attachments` like so

```ts
attachments: {
  columns: {
    id: true,
    // ...
  },
  orderBy(fields, { asc }) {
    return asc(fields.createdAt);
  }
}
```

## Behavior with `tsgo`
Not only it doesn't report any error, it even considers wrong API usage as correct and infers types. This would result in runtime error.

The type of `posts` is inferred as

```ts
const posts: {
  attachments: {
    createdAt: Date;
    filename: string | null;
    id: string;
    location: string;
    postId: string | null;
    size: number | null;
    type: string;
    updatedAt: Date;
  }[];
  id: string;
}[]
```

### Note
I have found out that the thing that breaks this is the `orderBy` function. If you comment that out, it will start reporting errors.

**Comment by RyanCavanaugh** [comment]
Copilot-reduced repro, verified:
```ts
type KnownKeysOnly<T, U> = {
    [K in keyof T]: K extends keyof U ? T[K] : never;
};

type Config = {
    child: { foo: unknown };
};

declare function test<T extends Config>(config: KnownKeysOnly<T, Config>): void;

test({
    child: {
        id: true,
        foo: 'x',
    }
});
```

**Comment by RyanCavanaugh** [comment]
Actually, this doesn't repro in 6.0 at latest...

**Comment by jansedlon** [comment]
Oh i'm so sorry, i'm on 5.9 actually

**Comment by RyanCavanaugh** [comment]
So was I 😅. Worth a bisect to see where this changed so that we can mention it in the 6.0 blog, investigating

**Comment by RyanCavanaugh** [comment]
Bisects to microsoft/TypeScript#62722

**Comment by RyanCavanaugh** [comment]
Well, that's not explicable. It seems intentional, though. @Andarist any thoughts?

**Comment by Andarist** [comment]
It seems to me the formulation of the given `KnownKeysOnly` is actually incorrect. It only tries only limit the *top-level* properties but the example call implicitly expects a property at a deeper level to be rejected. That really had to only work accidentally before my PR - given the inner object (a result of reverse mapped symbol inference) wasn't widened and thus it still was treated as a regular EPC target. 

In other words, in the given simplified example, it's not really `KnownKeysOnly` that leads to an error. It's the reverse mapped type itself. The same error gets reported if we switch it to using `Identity` ([TS 5.9 playground](https://www.typescriptlang.org/play/?ts=5.9.3#code/C4TwDgpgBAkgJhAdsAlqAPAFQHxQLxQDeAUFFANoDSUKiUA1hCAPYBmUmAugFwdWcBuYgF8hxUJCgBhZolYoA5viKkoAYwAWKADZxehKK2bNeAV0T1EzAO51RIsQjXaAhgCdorc2tSyowCABnYCwoCAAPAMQ4QOlZeQVsAAo1eMVeeCRUDBwASl4AN2YUODEA4KSSMk0dPRUyMhLeYDdTCAAaKAB6LrC3N2Y3QIA6UagAI1NgKGsNEAB+VTIjEygAInC19tVhbeFcoSA)):
```ts
type Identity<T> = {
  [K in keyof T]: T[K];
};

type Config = {
  child: { foo: unknown };
};

declare function test<T extends Config>(config: Identity<T>): void;

test({
  child: {
    id: true, // errors... but why?
    foo: "x",
  },
});
```

**Comment by RyanCavanaugh** [comment]
Thanks!

**Comment by jansedlon** [comment]
What does this mean for me exactly? Is this something DrizzleORM should fix on their side?

**Comment by Andarist** [comment]
This one seems to work ([TS 6.0 playground](https://www.typescriptlang.org/play/?ts=6.0.0-dev.20260226#code/C4TwDgpgBA0gdgewO5xhEBnA8nANiAHgBUAaKAVQD4oBeKAbwCgooBtGKASzigGt0EAMyhEAugC5YUCAA9gEOABMMfAcPJQA-CPaioAMliIUaTDnzFdZcruqS4EAG4QATgG5GAXw+NQkKADCCHCCnADmtAzMUADGABacuIqS9FCCCAiSAK5wvMY83l4+ihAxuACGLtCCOTHAnMFQ8hjAxNJyCsqBwaFhlAAUMT3hkvDIqOjYeISk3SHhlACUko4InIo+zcD9TAD0u1AAeprR8YnJUSws65LALlkQJNEs6ZlQAEQy708snk+ei02EBaO0Y+yOJxYZySKWeaQykk+32ify8gMYvmB2z2B2OpwSMMuV1eiK+PygqJYCGAcVcsKuUAARpVEa93ii0T4gA)):
```ts
type KnownKeysOnly<T, U> = {
  [K in keyof T]: K extends keyof U ? T[K] & KnownKeysOnly<T[K], U[K]> : never;
};
```
